content/en/blog/_posts/2025-02-14-cloud-controller-manager-chicken-egg-problem.md

```diff
@@ -0,0 +1,124 @@
+
+### Example: Cloud controller manager not scheduling due to not-ready taint
+
+The next example would be possible in situations where the container network interface (CNI) is waiting for IP address information from the cloud-controller-manager (CCM), and the CCM has not tolerated the taint which would be removed by the CNI.
+
+The Kubernetes documentation\[1\] describes the `node.kubernetes.io/not-ready` taint as follows:  
+“The Node controller detects whether a Node is ready by monitoring its health and adds or removes this taint accordingly.”  
+One of the conditions that can lead to a `Node` resource having this taint is when the container network has not yet been initialized on that node. As the cloud-controller-manager is responsible for adding the IP addresses to a `Node` resource, and the IP addresses are needed by the container network controllers to properly configure the container network, it is possible in some circumstances for a node to become stuck as not ready and uninitialized permanently.
+
+This situation occurs for a similar reason as the first example, although in this case, the `node.kubernetes.io/not-ready` taint is used with the no execute effect and thus will cause the cloud-controller-manager not to run on the node with the taint. If the cloud-controller-manager is not able to execute, then it will not initialize the node. It will cascade into the container network controllers not being able to run properly, and the node will end up carrying both the `node.cloudprovider.kubernetes.io/uninitialized` and `node.kubernetes.io/not-ready` taints, leaving the cluster in an unhealthy state.
+
+## Recommendations
+
+There is no one “correct way” to run a cloud-controller-manager. The details will depend on the specific needs of the cluster administrators and users. When planning your clusters and the lifecycle of the cloud-controller-managers please consider the following guidance:
+
+For cloud-controller-managers running in the same cluster, they are managing.
+
+1. Enable host networking. In most cases, a cloud controller manager will need to communicate with an API service endpoint associated with the infrastructure. Setting “hostNetwork” to true will ensure that the cloud controller is using the host networking instead of the container network and, as such, will have the same network access as the host operating system. It will also remove the dependency on the networking plugin. This will ensure that the cloud controller has access to the infrastructure endpoint (always check your networking configuration against your infrastructure provider’s instructions).  
```

Comment:
How I'd write it:
```suggestion
1. Use host network mode, rather than the pod network: in most cases, a cloud controller manager
   will need to communicate with an API service endpoint associated with your infrastructure.
   Setting `hostNetwork` to true ensures that the cloud controller manager is using the node's network interface
   instead of the container network and, as such, will have the same network access as the host operating system.
   Using host network mode also avoids a dependency on the cluster's network plugin(s), and that choice helps to ensure
   that the cloud controller manager can access to infrastructure-relevant API endpoints.  
   Always check your networking configuration against your infrastructure provider’s instructions.
```